types/src/v0/impls/reward.rs

```diff
@@ -0,0 +1,421 @@
+                    proof: RewardMerkleProof::Presence(proof),
+                },
+                balance.0,
+            )),
+            LookupResult::NotFound(proof) => Some((
+                Self {
+                    account,
+                    proof: RewardMerkleProof::Absence(proof),
+                },
+                0.into(),
+            )),
+            LookupResult::NotInMemory => None,
+        }
+    }
+
+    pub fn verify(&self, comm: &RewardMerkleCommitment) -> anyhow::Result<U256> {
+        match &self.proof {
+            RewardMerkleProof::Presence(proof) => {
+                ensure!(
+                    RewardMerkleTree::verify(comm.digest(), RewardAccount(self.account), proof)?
+                        .is_ok(),
+                    "invalid proof"
+                );
+                Ok(proof
+                    .elem()
+                    .context("presence proof is missing account balance")?
+                    .0)
+            },
+            RewardMerkleProof::Absence(proof) => {
+                let tree = RewardMerkleTree::from_commitment(comm);
+                ensure!(
+                    tree.non_membership_verify(RewardAccount(self.account), proof)?,
+                    "invalid proof"
+                );
+                Ok(0.into())
+            },
+        }
+    }
+
+    pub fn remember(&self, tree: &mut RewardMerkleTree) -> anyhow::Result<()> {
+        match &self.proof {
+            RewardMerkleProof::Presence(proof) => {
+                tree.remember(
+                    RewardAccount(self.account),
+                    proof
+                        .elem()
+                        .context("presence proof is missing account balance")?,
+                    proof,
+                )?;
+                Ok(())
+            },
+            RewardMerkleProof::Absence(proof) => {
+                tree.non_membership_remember(RewardAccount(self.account), proof)?;
+                Ok(())
+            },
+        }
+    }
+}
+
+impl From<(RewardAccountProof, U256)> for RewardAccountQueryData {
+    fn from((proof, balance): (RewardAccountProof, U256)) -> Self {
+        Self { balance, proof }
+    }
+}
+
+pub fn apply_rewards(
+    mut reward_state: RewardMerkleTree,
+    validator: Validator<BLSPubKey>,
+) -> anyhow::Result<RewardMerkleTree> {
+    let mut update_balance = |account: &RewardAccount, amount: RewardAmount| {
+        reward_state = reward_state.persistent_update_with(account, |balance| {
+            let balance = balance.copied();
+            match balance.unwrap_or_default().0.checked_add(amount.0) {
+                Some(updated) => Some(updated.into()),
+                None => {
+                    tracing::warn!("overflowed reward balance for account {}", account);
+                    balance
+                },
+            }
+        })?;
+        Ok::<(), anyhow::Error>(())
+    };
+
+    // Distribute leader reward
+    let leader_reward = RewardAmount::from(
```

Comment:
If the leader has no delegators does this calculation give the leader all the rewards?  I feel like this function is missing a piece where the leader should get rewards based on how much it is personally staking, not just the commissions.  Maybe this is accounted for because the leader is in the delegators?

Reply:
Yes, the leader will be among the delegators. I believe the way the contract is designed, the leader will be a delegator if it wants to stake

Reply:
I guess there should a check here that leader is among the delegators?